### PR TITLE
Make --build-shared-library more robust.

### PR DESCRIPTION
**Pull request by mraleph**
* Search for a suitable ARM sysroot instead of hardcoding it.

* Add facility to explain why NDK was not found.

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 1]
```diff
@@ -59,9 +59,122 @@ String getAdbPath([AndroidSdk existingSdk]) {
...
+    }
+
+    String findCompiler(String ndkDirectory) {
+      String directory;
+      if (platform.isLinux) {
+        directory = 'linux-x86_64';
+      } else if (platform.isMacOS) {
+        directory = 'darwin-x86_64';
+      } else {
+        throw new AndroidNdkSearchError('Only Linux and Mac OS are supported');
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: Ultra-nit: these days it's `macOS`, though they'll never convince me to stop calling it OS X ;-)

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 2]
```diff
@@ -379,7 +379,7 @@ Future<Null> _buildGradleProjectV2(String gradle, BuildInfo buildInfo, String ta
     if (buildInfo.fileSystemScheme != null)
       command.add('-Pfilesystem-scheme=${buildInfo.fileSystemScheme}');
   }
-  if (buildInfo.preferSharedLibrary && androidSdk.ndkCompiler != null) {
+  if (buildInfo.preferSharedLibrary && androidSdk.ndk != null) {
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Oops -- missed the gradle side yesterday. Once this lands I can make the naming updates.

**Review by cbracken — APPROVED** [review]
LGTM - looks great. Thanks for doing the cleanup in android_sdk.dart.

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: Done.

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 2 above)
Comment (mraleph) [reply packages/flutter_tools/lib/src/android/gradle.dart]: oops, I did not see that it says `preferSharedLibrary` either :) I guess this code is not covered with tets.

File: packages/flutter_tools/test/android/android_sdk_test.dart [hunk 3]
```diff
@@ -132,9 +132,11 @@ void main() {
 
           final AndroidSdk sdk = AndroidSdk.locateAndroidSdk();
           expect(sdk.directory, realSdkDir);
-          expect(sdk.ndkDirectory, realNdkDir);
-          expect(sdk.ndkCompiler, realNdkCompiler);
-          expect(sdk.ndkCompilerArgs, <String>['--sysroot', realNdkSysroot]);
+          print(AndroidNdk.explainMissingNdk(sdk.directory));
```
Comment (Hixie) [inline packages/flutter_tools/test/android/android_sdk_test.dart]: I'm assuming this `print` was unintentional. I've removed it in https://github.com/flutter/flutter/pull/17401/files; please let me know if that was a mistake.

Comment (mraleph) [reply packages/flutter_tools/test/android/android_sdk_test.dart]: Thanks, @Hixie. That was a debugging print left-over.

**Pull request merged**